File: hot100/图论/200.dfs.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    void dfs(vector<vector<char>>& grid, int r, int c) {
        int nr = grid.size();
        int nc = grid[0].size();

        if (r < 0 || c < 0 || r >= nr || c >= nc || grid[r][c] == '0') {
            return;
        }

        // 将当前位置标记为已访问
        grid[r][c] = '0';

        // 向四个方向递归
        dfs(grid, r - 1, c); // 上
        dfs(grid, r + 1, c); // 下
        dfs(grid, r, c - 1); // 左
        dfs(grid, r, c + 1); // 右
    }

    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty()) {
            return 0;
        }

        int nr = grid.size();
        int nc = grid[0].size();
        int num_islands = 0;

        for (int r = 0; r < nr; ++r) {
            for (int c = 0; c < nc; ++c) {
                if (grid[r][c] == '1') {
                    ++num_islands;
                    dfs(grid, r, c);
                }
            }
        }

        return num_islands;
    }
};

int main() {
    // 测试用例
    vector<vector<char>> grid1 = {
        {'1', '1', '1', '1', '0'},
        {'1', '1', '0', '1', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '0', '0', '0'}
    };

    vector<vector<char>> grid2 = {
        {'1', '1', '0', '0', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '1', '0', '0'},
        {'0', '0', '0', '1', '1'}
    };

    Solution solution;

    cout << "Number of islands in grid1: " << solution.numIslands(grid1) << endl;
    cout << "Number of islands in grid2: " << solution.numIslands(grid2) << endl;

    return 0;
}

File: hot100/图论/200.dfs.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(grid, i, j):
            # Check boundary conditions and whether the cell is water ('0')
            if not (0 <= i < len(grid)) or not (0 <= j < len(grid[0])) or grid[i][j] == '0':
                return
            # Mark the cell as visited
            grid[i][j] = '0'
            # Explore all 4 directions
            dfs(grid, i + 1, j)
            dfs(grid, i, j + 1)
            dfs(grid, i - 1, j)
            dfs(grid, i, j - 1)

        count = 0
        # Iterate through each cell in the grid
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                # If we find land ('1'), start a DFS to mark the entire island
                if grid[i][j] == '1':
                    dfs(grid, i, j)
                    count += 1  # Increment island count
        
        return count

# Structure to hold test case information
class TestCase:
    def __init__(self, grid: List[List[str]], expected: int):
        self.grid = grid
        self.expected = expected

# Function to run test cases
def runTests():
    # Create test cases
    testCases = [
        TestCase([
            ["1", "1", "0", "0", "0"],
            ["1", "1", "0", "0", "0"],
            ["0", "0", "1", "0", "0"],
            ["0", "0", "0", "1", "1"]
        ], 3),
        
        TestCase([
            ["1", "1", "1", "1", "0"],
            ["1", "1", "0", "1", "0"],
            ["1", "1", "0", "0", "0"],
            ["0", "0", "0", "0", "0"]
        ], 1),
        
        TestCase([
            ["1", "0", "0", "1"],
            ["0", "0", "1", "1"],
            ["0", "1", "1", "0"],
            ["1", "0", "0", "0"]
        ], 3),
        
        TestCase([
            ["0", "0", "0", "0"],
            ["0", "0", "0", "0"],
            ["0", "0", "0", "0"],
            ["0", "0", "0", "0"]
        ], 0),
        
        TestCase([
            ["1"]
        ], 1),
        
        TestCase([
            ["1", "1"],
            ["1", "1"]
        ], 1),
        
        TestCase([
            ["1", "0"],
            ["0", "1"]
        ], 2)
    ]

    sol = Solution()
    for i, testCase in enumerate(testCases):
        # Make a deep copy of the grid to preserve the original grid for each test case
        grid_copy = [row[:] for row in testCase.grid]
        result = sol.numIslands(grid_copy)
        print(f"Test Case {i + 1}:")
        print(f"Grid:")
        for row in testCase.grid:
            print(row)
        print(f"Expected Result: {testCase.expected}")
        print(f"Actual Result: {result}")
        print(f"Test {'Passed' if result == testCase.expected else 'Failed'}\n")

if __name__ == "__main__":
    runTests()
